plutus-tx/src/PlutusTx/AsData.hs

```diff
@@ -0,0 +1,148 @@
+      [d|
+        data Example a = Ex1 Integer | Ex2 a a
+      |]
+@
+becomes
+@
+    newtype Example a = Example BuiltinData
+      deriving newtype (ToData, FromData, UnsafeFromBuiltinData)
+
+    pattern Ex1 :: (ToData a, UnsafeFromData a) => Integer -> Example a
+    pattern Ex1 i = Example (unsafeDataAsConstr -> ((==) 0 -> True, [unsafeFromBuiltinData -> i]))
+      where Ex1 i = Example (mkConstr 0 [toBuiltinData i])
+
+    pattern Ex2 :: (ToData a, UnsafeFromData a) => a -> a -> Example a
+    pattern Ex2 a1 a2 = Example (unsafeDataAsConstr -> ((==) 1 -> True, [unsafeFromBuiltinData -> a1, unsafeFromBuiltinData -> a2]))
+      where Ex2 a1 a2 = Example (mkConstr 1 [toBuiltinData a1, toBuiltinData a2])
+
+    {-# COMPLETE Ex1, Ex2 #-}
+@
+-}
+asData :: TH.Q [TH.Dec] -> TH.Q [TH.Dec]
+asData decQ = do
+  decs <- decQ
+  outputDecs <- for decs asDataFor
+  pure $ concat outputDecs
+
+asDataFor :: TH.Dec -> TH.Q [TH.Dec]
+asDataFor dec = do
+  -- th-abstraction doesn't include deriving clauses, so we have to handle that hereth-abstraction doesn't include deriving clauses, so we have to handle that here
+  case dec of
+    TH.DataD _ _ _ _ _ deriv | not (null deriv) -> fail "asData: can't handle deriving clauses"
+    _                                           -> pure ()
+
+  di@(TH.DatatypeInfo{TH.datatypeVariant=dVariant, TH.datatypeCons=cons, TH.datatypeName=name, TH.datatypeVars=tTypeVars}) <- TH.normalizeDec dec
+
+  -- Other stuff is data families and so on
+  unless (dVariant == TH.Datatype) $ fail $ "asData: can't handle datatype variant " ++ show dVariant
+  -- a fresh name for the new datatype, but same lexically as the old one
+  cname <- TH.newName (show name)
+  -- The newtype declaration
+  let ntD =
+        let con = TH.NormalC cname [(TH.Bang TH.NoSourceUnpackedness TH.NoSourceStrictness, TH.ConT ''BuiltinData)]
+            ntDeriv ctx = TH.DerivClause (Just TH.NewtypeStrategy) ctx
+            deriv = ntDeriv [TH.ConT ''ToData, TH.ConT ''FromData, TH.ConT ''UnsafeFromData]
+        in TH.NewtypeD [] name tTypeVars Nothing con [deriv]
+
+  -- The pattern synonyms, one for each constructor
+  pats <- ifor cons $ \conIx (TH.ConstructorInfo{TH.constructorName=conName, TH.constructorFields=fields, TH.constructorVariant=cVariant}) -> do
+    -- If we have a record constructor, we need to reuse the names for the
+    -- matching part of the pattern synonym
+    fieldNames <- case cVariant of
+      TH.RecordConstructor names -> pure names
+      -- otherwise whatever
+      _                          -> ifor fields $ \fieldIx _ -> TH.newName $ "arg" ++ show fieldIx
+    let extractFieldNames = fieldNames
+    createFieldNames <- for fieldNames (TH.newName . show)
+    patSynArgs <- case cVariant of
+      TH.NormalConstructor   -> pure $ TH.PrefixPatSyn extractFieldNames
+      TH.RecordConstructor _ -> pure $ TH.RecordPatSyn extractFieldNames
+      TH.InfixConstructor    -> fail "asData: Can't handle infix data constructors"
```

Comment:
TH.InfixPatSyn?

Reply:
To some degree I was just being lazy: I can't remember the last time I saw an infix data constructor, and it's not quite as simple (since it always has exactly two binders).

Reply:
I added it anyway.